### PR TITLE
[Utils] Add new merge-release-pr.py script.

### PR DESCRIPTION
**Pull request by tru**
This script helps the release managers merge backport PR's.

It does the following things:

* Validate the PR, checks approval, target branch and many other things.
* Rebases the PR
* Checkout the PR locally
* Pushes the PR to the release branch
* Deletes the local branch

I have found the script very helpful to merge the PR's.

### CODE REVIEW

File: llvm/utils/release/merge-release-pr.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/env python3
+# ===-- merge-release-pr.py  ------------------------------------------------===#
+#
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# ===------------------------------------------------------------------------===#
+#
+# Helper script that will merge a Pull Request into a release branch. It will first
```
Comment (boomanaiden154) [inline llvm/utils/release/merge-release-pr.py:10]: Writing this out as a doc string would be more conventional for Python.

File: llvm/utils/release/merge-release-pr.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+    def run_gh(self, gh_cmd: str, args: List[str]) -> str:
+        cmd = ["gh", gh_cmd, "-Rllvm/llvm-project"] + args
+        p = subprocess.run(cmd, capture_output=True)
+        if p.returncode != 0:
+            print(p.stderr)
+            raise RuntimeError("Failed to run gh")
+        return p.stdout
+
+    # Validate the state of the PR, this means making sure that it is
```
Comment (boomanaiden154) [inline llvm/utils/release/merge-release-pr.py:42]: Doc string here.

File: llvm/utils/release/merge-release-pr.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    # OPEN and not already merged or closed.
+    def validate_state(self, data):
+        state = data["state"]
+        if state != "OPEN":
+            return False, f"state is {state.lower()}, not open"
+        return True
+
+    # Validate that the PR is targetting a release/ branch. We could
```
Comment (boomanaiden154) [inline llvm/utils/release/merge-release-pr.py:50]: Docstring here.

File: llvm/utils/release/merge-release-pr.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    # validate the exact branch here, but I am not sure how to figure
+    # out what we want except an argument and that might be a bit to
+    # to much overhead.
+    def validate_target_branch(self, data):
+        baseRefName: str = data["baseRefName"]
+        if not baseRefName.startswith("release/"):
+            return False, f"target branch is {baseRefName}, not a release branch"
+        return True
+
+    # Validate the approval decision. This checks that the PR has been
```
Comment (boomanaiden154) [inline llvm/utils/release/merge-release-pr.py:60]: Docstring here.

File: llvm/utils/release/merge-release-pr.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    # approved.
+    def validate_approval(self, data):
+        if data["reviewDecision"] != "APPROVED":
+            return False, "PR is not approved"
+        return True
+
+    # Check that all the actions / status checks succeeded. Will also
```
Comment (boomanaiden154) [inline llvm/utils/release/merge-release-pr.py:67]: Docstring here.

File: llvm/utils/release/merge-release-pr.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                if failures:
+                    errstr += "\n"
+                errstr += "    PENDING: "
+                errstr += ", ".join([d["name"] for d in pending])
+
+            return False, errstr
+
+        return True
+
+    # Validate that the PR contains just one commit. If it has more
```
Comment (boomanaiden154) [inline llvm/utils/release/merge-release-pr.py:93]: Docstring here.

File: llvm/utils/release/merge-release-pr.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+import subprocess
+import sys
+import time
+from typing import List
+
+
+class PRMerger:
+    def __init__(self, args):
+        self.pr = args.pr
+        self.args = args
```
Comment (boomanaiden154) [inline llvm/utils/release/merge-release-pr.py:32]: It seems like we're including duplicate information here?

**Review by boomanaiden154 — COMMENTED** [review]
Main comment is about using docstrings over just comments about the function (https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings). We don't have an official Python style guide (maybe we should start one), but I think it's a reasonable enough format and standardized enough, at least in the code that I read (which is a _bit_ biased towards one style).

We could also use the Github API here to grab the PR info/do stuff, but using the `gh` tool seems good enough, and if that's what the release managers have setup, then that's what we should go with because they're the ones ultimately using it.

Writing tests looks like it would mostly be integration tests while mocking the Github API/CLI tool, which would be a massive pain for not much benefit.

**Review by tstellar — APPROVED** [review]
I think this looks good.  In theory, we might be able to add a workflow in the future to run this script automatically when a release manager approves a PR.

**Review by boomanaiden154 — APPROVED** [review]
LGTM.

**Pull request merged**